### ISSUE TITLE
Selection of Optimal Learning Rate for Adam Optimizer

### ISSUE THREAD

**Issue by rubeea** [open-mmlab/mmsegmentation#346]
Hi,
Can anyone suggest me how can one select an optimal learning rate while implementing the Adam optimizer. While using PSP decode head with ResNetv1c backbone the optimizer gives good results for LR=5e-5 but for a model with U-Net backbone and PSP decode Head is not learning at this LR at all. Any suggestion on how to suggest the optimal LR according to the model employed and how to implement a LR Scheduler to select the optimal learning rate.

Thanks

**Comment by xvjiarui** [comment]
The learning rate would be different across models/datasets. Personally, I will search it a little bit.

**Comment by rubeea** [comment]
> The learning rate would be different across models/datasets. Personally, I will search it a little bit.

Does mmsegmentation allow us to use a hyperparameter search framework such as optuna or Lr finder? Because it is very difficult to search the LR manually. Also can you explain what is the difference between the decode loss and the auxiliary loss? Also can be use different losses for these two heads?

**Comment by xvjiarui** [comment]
Hi @rubeea 
Hyperparameter search framework is not supported yet. 
The auxiliary head helps backbone learn a better representation, which could be dropped during inference. And their loss could be different.

**Comment by rubeea** [comment]
Hi @xvjiarui Thanks for your response :)

**Comment by rubeea** [comment]
Hi @xvjiarui, one last question. As far as I understand, I believe that decode and aux heads are being used in parallel in the model config right? Or are they cascaded i.e. aux head takes in the input of the decode head? 
Thanks in advance.

**Comment by xDooooot** [comment]
@rubeea, there is a function using in the training procedure, which named `forward_train` in `/path/to/mmsegmentation/mmseg/models/segmentors/encoder_decoder`, it seems that both decode head and aux head take the same inputs and produce different losses for backpropagation.

**Comment by rubeea** [comment]
> @rubeea, there is a function using in the training procedure, which named `forward_train` in `/path/to/mmsegmentation/mmseg/models/segmentors/encoder_decoder`, it seems that both decode head and aux head take the same inputs and produce different losses for backpropagation.

Hi,
Thanks for your reply but I figured out the difference. The aux and decode heads are indeed used in parallel but they take features of the decoder output as their input. Aux head takes the 3rd decoder output features while the decode head takes the last decoder output features for calculating the losses respectively. This can be seen in the 'in_index' field in the config fie.

**Comment by xDooooot** [comment]
> > @rubeea, there is a function using in the training procedure, which named `forward_train` in `/path/to/mmsegmentation/mmseg/models/segmentors/encoder_decoder`, it seems that both decode head and aux head take the same inputs and produce different losses for backpropagation.
> 
> Hi,
> Thanks for your reply but I figured out the difference. The aux and decode heads are indeed used in parallel but they take features of the decoder output as their input. Aux head takes the 3rd decoder output features while the decode head takes the last decoder output features for calculating the losses respectively. This can be seen in the 'in_index' field in the config file.

Indeed as you say, there could be different 'in_index' in the config file. Thanks for correcting me :)

**Comment by rubeea** [comment]
@xDooooot No worries. Have a nice day :)
